## Implement integer exponentiation.
This problem was asked by Google.

Implement integer exponentiation. That is, implement the pow(x, y) function, where x and y are integers and returns x^y.
Do this faster than the naive method of repeated multiplication.
For example, pow(2, 10) should return 1024.

In [5]:
def pow(n, m):
    return n**m

print(pow(2, 10))
pow(4, 3)

1024


64

For above solution:- Using the `**` operator is a concise and efficient way to calculate exponentiation in Python.However, the question also mentions doing this "faster than the naive method of repeated multiplication." The above solution using ** is likely more efficient than the naive method, especially for larger exponents, but it does not explicitly demonstrate the use of more efficient algorithms like binary exponentiation.

You can implement integer exponentiation more efficiently using the binary exponentiation technique. This approach reduces the number of multiplication operations required compared to the naive method of repeated multiplication.

In [7]:
# binary exponentiation technique. 

def pow(x,y):
    result = 1
    
    # Handle negative exponents
    if y < 0:
        x = 1/x
        y = -y
        
    # binary exponentiation algorithm
    while y > 0:
        if y % 2 == 1:
            result *= x
        x *= x
        y //= 2
    return result

# Example
print(pow(2, 5))
print(pow(2, -3))
print(pow(3, -2)) 


32
0.125
0.1111111111111111


### Explanation:

1. Initialize result to 1.
2. If y is negative, handle it by making x the reciprocal of itself and making y positive.
3. Iterate through each bit of y using binary exponentiation algorithm:
    1. If the least significant bit of y is 1, multiply result by x.
    2. Square x.
    3. Shift y to the right (equivalent to integer division by 2).
4. Return the final result.

```
print(pow(2, -3))  # Output: 0.125 (which is 1 / 2^3)
print(pow(3, -2))  # Output: 0.1111111111111111 (which is approximately 1 / 3^2)
```
In these examples, `pow(2, -3)` calculates `2` raised to the power `-3`, which is equivalent to `1 / 2^3 (or 1 / 8)`, resulting in `0.125`. Similarly, `pow(3, -2)` calculates `3` raised to the power `-2`, which is equivalent to `1 / 3^2 (or 1 / 9)`, resulting in `0.1111111111111111`.